enrollment-server-onboarding/src/main/java/com/wultra/app/onboardingserver/impl/service/IdentityVerificationPrecompleteGuard.java

```diff
@@ -0,0 +1,173 @@
+import com.wultra.app.onboardingserver.common.database.entity.DocumentVerificationEntity;
+import com.wultra.app.onboardingserver.common.database.entity.IdentityVerificationEntity;
+import com.wultra.app.onboardingserver.common.database.entity.OnboardingOtpEntity;
+import com.wultra.app.onboardingserver.common.errorhandling.RemoteCommunicationException;
+import com.wultra.app.onboardingserver.configuration.IdentityVerificationConfig;
+import com.wultra.app.onboardingserver.statemachine.guard.document.RequiredDocumentTypesGuard;
+import com.wultra.security.powerauth.client.v3.ActivationStatus;
+import lombok.Builder;
+import lombok.Getter;
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.stereotype.Component;
+
+import java.util.List;
+import java.util.stream.Collectors;
+
+import static com.wultra.app.enrollmentserver.model.enumeration.IdentityVerificationPhase.*;
+import static com.wultra.app.enrollmentserver.model.enumeration.IdentityVerificationStatus.ACCEPTED;
+import static com.wultra.app.enrollmentserver.model.enumeration.IdentityVerificationStatus.VERIFICATION_PENDING;
+
+/**
+ * Validate all critical conditions were met before finishing the onboarding.
+ * <p>
+ * This should never happen for the state machine.
+ * It works as a safety stop.
+ *
+ * @author Lubos Racansky, lubos.racansky@wultra.com
+ */
+// TODO (racansky, 2022-10-14) consider make it Guard for Spring State Machine
+@Component
+@Slf4j
+class IdentityVerificationPrecompleteGuard {
+
+    private final IdentityVerificationConfig identityVerificationConfig;
+
+    private final RequiredDocumentTypesGuard requiredDocumentTypesGuard;
+
+    private final OnboardingOtpRepository onboardingOtpRepository;
+
+    private final ActivationService activationService;
+
+    IdentityVerificationPrecompleteGuard(
+            final IdentityVerificationConfig identityVerificationConfig,
+            final RequiredDocumentTypesGuard requiredDocumentTypesGuard,
+            final OnboardingOtpRepository onboardingOtpRepository,
+            final ActivationService activationService) {
+        this.identityVerificationConfig = identityVerificationConfig;
+        this.requiredDocumentTypesGuard = requiredDocumentTypesGuard;
+        this.onboardingOtpRepository = onboardingOtpRepository;
+        this.activationService = activationService;
+    }
+
+    /**
+     * Evaluate all precomplete conditions.
+     *
+     * @param idVerification identity verification to evaluate
+     * @return evaluation result
+     */
+    Result evaluate(final IdentityVerificationEntity idVerification) throws RemoteCommunicationException {
+        final List<DocumentVerificationEntity> documentVerifications = idVerification.getDocumentVerifications().stream()
+                .filter(it -> DocumentStatus.ALL_PROCESSED.contains(it.getStatus()))
+                .collect(Collectors.toList());
+
+        if (!documentVerifications.stream()
```

Comment:
I am not sure this check is correct. A document can be `REJECTED` or `FAILED` and later `ACCEPTED` after reupload as another DB row.

We should rather check that 2 required documents were `ACCEPTED` and these documents have `usedForVerification = true`, no matter the status of the remaining documents.

Reply:
👀  Good catch. Logic with `REJECTED`, `FAILED`, and `ACCEPTED` is correct. But I have optimized it too much. There used to be
```java
List<DocumentVerificationEntity> processedDocVerifications =
                documentVerificationRepository.findAllDocumentVerifications(idVerification, DOCUMENT_STATUSES_PROCESSED);
```
which I replaced by
```java
idVerification.getDocumentVerifications().stream()
                .filter(it -> DocumentStatus.ALL_PROCESSED.contains(it.getStatus()))
                .collect(Collectors.toList());
```
It gives different result e.g. for `DISPOSED` representing resubmitted documents.

Reply:
👍